[expo-updates] add unit tests for manifest classes

esamelson opened the pull request:
# Why

Slowly but surely improving the automated test coverage of this module!

# How

Added native unit tests for the following things:
- legacy manifest required fields
- new (EAS) manifest required fields
- manifest creation (that the correct class is used depending on the type of manifest)
- added a bit more coverage to the existing tests around asset base urls

Additionally, writing these tests exposed some discrepancies between/within platforms about handling possibly null manifest fields, so I fixed those. Thanks, tests!

# Test Plan

All tests pass ✅ 

# To do

- setup a CI job for Xcode unit tests

esamelson committed:
[android] add native unit tests around manifest types and required fields

esamelson committed:
[ios] add native unit tests around manifest types and required fields, correct handling of manifest fields

esamelson committed:
[ios] fix handling of nil assets array

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/ManifestFactoryTest.java:
```diff
@@ -0,0 +1,83 @@
+public class ManifestFactoryTest {
+
+  private String legacyManifestJson = "{\"sdkVersion\":\"39.0.0\",\"id\":\"@esamelson/native-component-list\",\"releaseId\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"commitTime\":\"2020-11-11T00:17:54.797Z\",\"bundleUrl\":\"https://d1wp6m56sqw74a.cloudfront.net/%40esamelson%2Fnative-component-list%2F39.0.0%2F01c86fd863cfee878068eebd40f165df-39.0.0-ios.js\"}";
+  private String newManifestJson = "{\"manifest\":{\"runtimeVersion\":\"1\",\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"createdAt\":\"2020-11-11T00:17:54.797Z\",\"launchAsset\":{\"url\":\"https://d1wp6m56sqw74a.cloudfront.net/%40esamelson%2Fnative-component-list%2F39.0.0%2F01c86fd863cfee878068eebd40f165df-39.0.0-ios.js\",\"contentType\":\"js\"}}}";
```
```suggestion
  private String newManifestJson = "{\"manifest\":{\"runtimeVersion\":\"1\",\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"createdAt\":\"2020-11-11T00:17:54.797Z\",\"launchAsset\":{\"url\":\"https://d1wp6m56sqw74a.cloudfront.net/%40esamelson%2Fnative-component-list%2F39.0.0%2F01c86fd863cfee878068eebd40f165df-39.0.0-ios.js\",\"contentType\":\"application/javascript\"}}}";
```
this is more realistic

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/NewManifestTest.java:
```diff
@@ -0,0 +1,70 @@
+  @Test
+  public void testFromManifestJson_AllFields() throws JSONException {
+    // production manifests should require the id, createdAt, runtimeVersion, and launchAsset fields
+    String manifestJson = "{\"runtimeVersion\":\"1\",\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"createdAt\":\"2020-11-11T00:17:54.797Z\",\"launchAsset\":{\"url\":\"https://url.to/bundle.js\",\"contentType\":\"js\"}}";
```
```suggestion
    String manifestJson = "{\"runtimeVersion\":\"1\",\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"createdAt\":\"2020-11-11T00:17:54.797Z\",\"launchAsset\":{\"url\":\"https://url.to/bundle.js\",\"contentType\":\"application/javascript\"}}";
```

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/NewManifestTest.java:
```diff
@@ -0,0 +1,70 @@
+    Assert.assertNotNull(NewManifest.fromManifestJson(manifest, createConfig()));
+  }
+
+  @Test(expected = Exception.class)
```
Is there a more precise type of Exception that's always expected?

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/manifest/NewManifestTest.java:
```diff
@@ -0,0 +1,70 @@
+
+  @Test
+  public void testFromManifestJson_StripsOptionalRootLevelKeys() throws JSONException {
+    String manifestJsonWithRootLevelKeys = "{\"data\":{\"publicManifest\":{\"manifest\":{\"runtimeVersion\":\"1\",\"id\":\"0eef8214-4833-4089-9dff-b4138a14f196\",\"createdAt\":\"2020-11-11T00:17:54.797Z\",\"launchAsset\":{\"url\":\"https://url.to/bundle.js\",\"contentType\":\"js\"}}}}}";
```
Conflating this with a JSON manifest seems like it will cause confusion later on. This is an EAS-specific API server response.

ide replied:
Talking with Eric -- we're going to make the change on the server first before changing the client-side code. This PR just adds tests for code that's already there.

ide commented on packages/expo-updates/ios/Tests/EXUpdatesNewUpdateTests.m:
```diff
@@ -0,0 +1,108 @@
+    @"runtimeVersion": @"1",
+    @"id": @"0eef8214-4833-4089-9dff-b4138a14f196",
+    @"createdAt": @"2020-11-11T00:17:54.797Z",
+    @"launchAsset": @{@"url": @"https://url.to/bundle.js", @"contentType": @"js"}
```
```suggestion
    @"launchAsset": @{@"url": @"https://url.to/bundle.js", @"contentType": @"application/javascript"}
```

ide reviewed (changes requested):
Would be good to add tests with an origin-relative override like `/my_assets` too.

esamelson committed:
fix contentType and add specific expected exception

esamelson commented:
addressed all code comments; re: origin-relative URLs, it looks like we don't actually handle those correctly right now (!) . I will open a new PR (with tests) to address that rather than doing it all in this PR.

esamelson committed:
update changelog
